### PR TITLE
Implement `FromStr`

### PR DESCRIPTION
**Pull request by TonalidadeHidrica**
I implemented `FromStr`.  `TryFromIntError` was renamed to a more semantic name `OutOfRangeError`.

### CODE REVIEW

**Comment by jhpratt** [comment]
Implementing `FromStr` is reasonable, but I'd like to see errors follow the same pattern as in `std`. Namely creating a `ParseIntError` with the same API as [`std::num::ParseIntError`](https://doc.rust-lang.org/stable/std/num/struct.ParseIntError.html) and having it use the public [`std::num::IntErrorKind`](https://doc.rust-lang.org/stable/std/num/enum.IntErrorKind.html).

The general purpose for this library is to serve as a playground before ranged integers are proposed in Rust itself, which is why I'm trying to provide APIs nearly identical to that of `std`.

**Comment by TonalidadeHidrica** [comment]
@jhpratt Then should we not introduce a new `OutOfRangeError` but reuse `PosOverflow` and `NegOverflow` for out-of-bounds input?

**Comment by jhpratt** [comment]
Correct. It fits with the documentation of "Integer is too large/small to store in target integer type."

**Comment by TonalidadeHidrica** [comment]
@jhpratt Would you mind adding a field to `TryFromIntError` describing whether it is a positive or negative overflow?

**Comment by jhpratt** [comment]
`TryFromIntError` should be unchanged. A new error type needs to be created.

**Comment by TonalidadeHidrica** [comment]
@jhpratt I think I fixed `ParseIntError` as you described.  Why I asked about adding a field to `TryFromIntError` was that I tried not to re-implement the boundary checking conditional branch but reuse that of `TryFrom` implementation instead.  But in `FromStr` I needed an information of whether it exceeds or falls behind.  If `TryFromIntError` contained the information, we could reuse it, but you said we shouldn't so I kept it untouched.

When it comes to adding a new error types, we may add a new `enum OutOfBoundsError { GreaterThanMax, LessThanMin }` thing and let `U32::new` return `Result<U32, OutOfBoundsError>` instead of `Option<U32>`.  This possible improvement can be done in a separate PR, whether or not you agree with this suggestion, anyway.

**Comment by jhpratt** [comment]
Having the bounds check in multiple places isn't an issue by itself. [`TryFromIntError` has no data](https://doc.rust-lang.org/stable/src/core/num/error.rs.html#9) in `std`, which is why it doesn't here either.

`U32::new` is based on [`NonZeroU32::new`](https://doc.rust-lang.org/stable/std/num/struct.NonZeroU32.html#method.new), which returns `Option<NonZeroU32>`. Like I said, my goal is to mimic the API in `std` as it already exists, so I'd prefer not to introduce a new error type that isn't present there.

**Review by jhpratt — COMMENTED** [review]
One final thing, I promise!

**Comment by TonalidadeHidrica** [comment]
@jhpratt Sorry for disrupting, I did understand that the purpose is to match the API to what is already in stdlib, so I didn't mean to insisting on adding a field anymore.  I also understood that rewritten bounds check does not matter, so I think this PR is ready to go.

File: src/lib.rs [hunk 1]
```diff
@@ -52,6 +54,41 @@ impl fmt::Display for TryFromIntError {
...
 impl Error for TryFromIntError {}
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ParseIntError {
+    kind: IntErrorKind,
+}
+
+impl ParseIntError {
+    /// Outputs the detailed cause of parsing an integer failing.
+    pub const fn kind(&self) -> &IntErrorKind {
```
[the remark this answers is not in the document]
Comment (TonalidadeHidrica) [reply src/lib.rs:64]: @jhpratt No it isn't.  I added because `clippy` asked me to do so, but should I just `allow` it instead?

**Comment by jhpratt** [comment]
Not an issue 🙂

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jhpratt) [reply src/lib.rs:64]: Yeah, let's do that.

**Comment by jhpratt** [comment]
Thanks!

**Pull request merged**